File: C++/atdc_sort.cpp
#ifndef atdc_sort

#include <iostream>
#include <iterator>
#include <vector>
#include <algorithm>
#include <stdexcept>
#include <functional>

#define atdcSort atdc_sort
#define atdcsort atdc_sort
#define AtdcSort atdc_sort
#define Atdcsort atdc_sort

template <typename RAI, typename X>
void __merge (RAI b1, RAI e1, RAI b2, RAI e2, X comp)
{
	std::vector <typename RAI::value_type> __arr;
	RAI it1 = b1, it2 = b2;
	while (it1 != e1 && it2 != e2)
	{
		if (comp (*it1, *it2) == 1)
		{
			__arr.push_back (*it1);
			it1 ++;
		}
		else if (comp (*it1, *it2) == 2)
		{
			__arr.push_back (*it2);
			it2 ++;
		}
	}
	
	if (it1 == e1)
	{
		while (it2 != e2)
		{
			__arr.push_back (*it2);
			it2 ++;
		}
	}
	else if (it2 == e2)
	{
		while (it1 != e1)
		{
			__arr.push_back (*it1);
			it1 ++;
		}
	}
	
	it1 = b1;
	for (size_t i = 0 ; i < (size_t)distance (b1, e1) ; i ++)
	{
		*it1 = __arr [i];
		it1 ++;
	}
	
	it2 = b2;
	for (size_t i = (size_t)distance (b1, e1) ; i < __arr.size () ; i ++)
	{
		*it2 = __arr [i];
		it2 ++;
	}
}

template <typename RAI, typename X = std::less<typename RAI::value_type> >
void __atdc_sort (RAI Begin, RAI End, X comp)
{
	int dist = std::distance (Begin, End);
	if (dist == 0)
		return ;
	if (dist == 1)
		return ;
	if (dist == 2)
	{
		RAI last = prev (End);
		typename RAI::value_type b = *Begin;
		if (comp (*Begin, *last) == 2)
		{
			*Begin = *last;
			*last = b;
		}
		return ;
	}
	
	dist /= 2;
	RAI mid = Begin + dist;
	__atdc_sort (Begin, mid, comp);
	__atdc_sort (mid, End, comp);
	__merge (Begin, mid, mid, End, comp);
}

template <typename RAI, typename X = std::less<typename RAI::value_type> >
void atdc_sort (RAI Begin, RAI End, X comp = X())
{
	__atdc_sort (Begin, End, comp);
}
#endif

File: C#/MONO_version/atdc_sort/atdc_sort.cs
namespace AlexRo6afF
{
	using System;
	using System.Collections.Generic;

	public delegate int Comp<T>(T a, T b);

	public static class ATDC
	{
		private static void Merge(dynamic what, int b1, int e1, int b2, int e2, Comp<dynamic> comp = null)
		{
			if (comp==null)
			{
				comp = (a, b) => { return a.CompareTo(b); };
			}

			List<dynamic> __arr = new List<dynamic>();
			int it1 = b1, it2 = b2;
			while (it1 != e1 && it2 != e2)
			{
				if (comp(what[it1], what[it2]) == -1)
				{
					__arr.Add(what[it1]);
					it1++;
				}
				else
				{
					__arr.Add(what[it2]);
					it2++;
				}
			}

			if (it1 == e1)
			{
				while (it2 != e2)
				{
					__arr.Add(what[it2]);
					it2++;
				}
			}

			if (it2 == e2)
			{
				while (it1 != e1)
				{
					__arr.Add(what[it1]);
					it1++;
				}
			}

			it1 = b1;
			for (int i = 0; i < e1 - b1; i++)
			{
				what[it1] = __arr[i];
				it1++;
			}

			it2 = b2;
			for (int i = e1 - b1; i < __arr.Count; i++)
			{
				what[it2] = __arr[i];
				it2++;
			}
		}

		public static void Sort(dynamic what, int begin, int end, Comp<dynamic> comp = null)
		{
			if (comp == null)
			{
				comp = (a, b) => { return a.CompareTo(b); };
			}
			int last = end - 1;
			int dist = end - begin + 1;
			if (dist == 0 || dist == 1)
			{
				return;
			}
			if (dist == 2)
			{
				if (comp(what[begin], what[last]) == -1)
				{
					int tmp = what[begin];
					what[begin] = what[end];
					what[last] = tmp;
				}
				return;
			}

			dist /= 2;
			int mid = begin + dist;
			Sort(what, begin, mid, comp);
			Sort(what, mid, end, comp);
			Merge(what, begin, mid, mid, end, comp);
		}
	}
}